### PR TITLE
Don't warn CA1802 when type's name matches enum's name

### LINKED ISSUE #6681: False Positive for CA1802 when using readonly with enum Flags
### Analyzer

**Diagnostic ID**: [CA1802](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/CA1802): `Use Literals Where Appropriate`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 7.0.1

### Describe the bug

CA1802: Field 'BindingFlags' is declared as 'readonly' but is initialized with a constant value. Mark this field as 'const' instead.
But this does not work for `enum`

### Steps To Reproduce

* New class
* Add `using System.Reflection;`
* Add `private static readonly BindingFlags BindingFlags = BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance;`
* See error: CA1802: Field 'BindingFlags' is declared as 'readonly' but is initialized with a constant value. Mark this field as 'const' instead.

### Expected behavior
No Error/Warning for CA1802

### Actual behavior
CA1802: Field 'BindingFlags' is declared as 'readonly' but is initialized with a constant value. Mark this field as 'const' instead.

### PR DESCRIPTION
**Pull request by CollinAlpert**
Currently the analyzer suggests to change a `static readonly` field declaration to a `const` if possible. This results in a compilation error if the field has the same name as the type.

Fixes #6681.

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/UseLiteralsWhereAppropriate.cs [hunk 1]
```diff
@@ -70,6 +70,7 @@ public override void Initialize(AnalysisContext context)
                         lastField == null ||
                         lastField.IsConst ||
                         !lastField.IsReadOnly ||
+                        lastField.Type.Name == lastField.Name ||
```
[the remark this answers is not in the document]
Comment (CollinAlpert) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/UseLiteralsWhereAppropriate.cs:73]: Yes, true. What do you think about extending the condition to the following:
```
lastField.Type.TypeKind == TypeKind.Enum && lastField.Type.Name == lastField.Name && fieldInitializerValue is IBinaryOperation or IFieldReferenceOperation
```

[the remark this answers is not in the document]
Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/UseLiteralsWhereAppropriate.cs:73]: > private const BindingFlags BindingFlags = (BindingFlags)(BindingFlags.Public | BindingFlags.NonPublic)

This still produces the error, and the right side is an `IConversionOperation`. I am fine with the simple check you have in the PR. If you want to make it stricter, the correct check would be to verify that the IOperation tree rooted at fieldInitializer does not contain any `IFieldReferenceOperation` whose field symbol's containing type is same as `lastField.Type`. Either way works for me - please do add both the test cases mentioned in this comment thread to the new unit test added in the PR.

[the remark this answers is not in the document]
Comment (CollinAlpert) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/UseLiteralsWhereAppropriate.cs]: Done. I opted for the more extensive check.

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/UseLiteralsWhereAppropriate.cs [hunk 2]
```diff
@@ -72,6 +72,7 @@ public override void Initialize(AnalysisContext context)
                         lastField == null ||
                         lastField.IsConst ||
                         !lastField.IsReadOnly ||
+                        (lastField.Type.Name == lastField.Name && fieldInitializerValue.DescendantsAndSelf().Any(d => d is IFieldReferenceOperation field && field.Field.Type.Equals(lastField.Type, SymbolEqualityComparer.Default))) ||
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/UseLiteralsWhereAppropriate.cs:75]: I always forget if/when the `Type` can be null for field symbols and field reference operations. A more defensive check would be:
```suggestion
                        (lastField.Type?.Name == lastField.Name && fieldInitializerValue.DescendantsAndSelf().Any(d => d is IFieldReferenceOperation field && lastField.Type.Equals(field.Field.Type, SymbolEqualityComparer.Default))) ||
```

Comment (CollinAlpert) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/UseLiteralsWhereAppropriate.cs]: Good idea. Done.

**Review by mavasani — APPROVED** [review]

**Pull request merged**